### PR TITLE
Limit ResizeProcessor memory consumption

### LINKED ISSUE #642: Optimize memory consumption of ResizeProcessor
Currently we are using `O(NewWidth * SourceHeight)` memory for doing resize operations, however it's possible to do it consuming only `O(NewWidth * KernelDiameter)` memory, without significant negative effect on speed.

We need to implement a "sliding window" algorithm for this.

It will most likely fix issues like #597.

**UPDATE**
~The issue is now part of the epic #733.~

**UPDATE 2**
The ideas in #733 are abandoned now, we need to implement this with the current architecture.

### PR DESCRIPTION
**Pull request by antonfirsov**
### Prerequisites
- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

## Description
This PR fixes #642 by limiting the memory consumption of `ResizeProcessor` to a value defined in `Configuration.WorkingBufferSizeHintInBytes`. Instead of allocating the whole first pass buffer, a sliding window is implemented by `ResizeWorker<T>`. The height of the window is a multiple of the vertical kernel's maximum diameter depending on `Configuration.WorkingBufferSizeHintInBytes`.

The implementation is most likely very different from MagicScaler's one (haven't seen it's code). It was a goal to keep the changes under control, the new code is basically a refactor of the old one.

The optimization is visually explained in **[ResizeWorker.pptx](https://github.com/SixLabors/ImageSharp/blob/af/resize-sandbox/src/ImageSharp/Processing/Processors/Transforms/Resize/ResizeWorker.pptx)**:

image

image

image

## Results
There is no regression in speed. (Actually, [we are faster now for .NET 4.72](https://github.com/SixLabors/ImageSharp/blob/af/resize-sandbox/tests/ImageSharp.Benchmarks/Samplers/Resize.cs#L120) than before.)

The resizing of a `4000 x 4000` image to `300 x 300` has been memory profiled using JetBrains dotMemory. The resize operation has been executed 50 times in a single process.

### Memory profile before the optimization
Timeline:
image
image

Typical snapshot:
image

### Memory profile after the optimization
Timeline:
image

image

(The second peak has been only triggered because I made the snapshot. Can anyone explain this?)

Typical snapshot:
image

### CODE REVIEW

**Comment by iamcarbon** [comment]
@antonfirsov  -- BTW: the new travis build error is fixed in my C# 8.0 PR.

File: tests/ImageSharp.Benchmarks/Samplers/Resize.cs [hunk 1]
```diff
@@ -94,26 +101,44 @@ protected override void ExecuteResizeOperation(IImageProcessingContext<Rgba32> c
...
-        //  'ImageSharp, MaxDegreeOfParallelism = 1' | Core |    Core |       3032 |      400 |  90.28 ms |  4.679 ms | 0.2565 ms |  0.76 |           - |           - |           - |             15712 B |
+        //                                    Method |  Job | Runtime | SourceToDest |      Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
+        // ----------------------------------------- |----- |-------- |------------- |----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
+        //                             SystemDrawing |  Clr |     Clr |     3032-400 | 120.11 ms |  1.435 ms | 0.0786 ms |  1.00 |    0.00 |           - |           - |           - |              1638 B |
+        //  'ImageSharp, MaxDegreeOfParallelism = 1' |  Clr |     Clr |     3032-400 |  75.32 ms | 34.143 ms | 1.8715 ms |  0.63 |    0.02 |           - |           - |           - |             16384 B |
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Benchmarks/Samplers/Resize.cs]: .NET Framework is faster?

File: tests/ImageSharp.Tests/Memory/Buffer2DTests.cs [hunk 2]
```diff
@@ -127,5 +127,56 @@ public void SwapOrCopyContent()
...
         }
+
+        [Theory]
+        [InlineData(100, 20, 0, 90, 10)]
+        [InlineData(100, 3, 0, 50, 50)]
+        [InlineData(123, 23, 10, 80, 13)]
+        [InlineData(10, 1, 3, 6, 3)]
+        [InlineData(2, 2, 0, 1, 1)]
+        [InlineData(5, 1, 1, 3, 2)]
+        public void CopyColumns(int width, int height, int startIndex, int destIndex, int columnCount)
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Memory/Buffer2DTests.cs]: Good tests!

File: tests/ImageSharp.Benchmarks/Samplers/Resize.cs (hunk 1 above)
Comment (antonfirsov) [reply tests/ImageSharp.Benchmarks/Samplers/Resize.cs]: I'm also surprised, but happens quite often when you go full unsafe, dropping all the span indexers. It is also backed by RyuJIT.

File: src/ImageSharp/Memory/Buffer2DExtensions.cs [hunk 3]
```diff
@@ -14,55 +17,135 @@ namespace SixLabors.ImageSharp.Memory
...
+
+            fixed (byte* ptr = span)
+            {
+                byte* basePtr = (byte*)ptr;
+                for (int y = 0; y < buffer.Height; y++)
+                {
+                    byte* sPtr = basePtr + sOffset;
+                    byte* dPtr = basePtr + dOffset;
+
+                    Buffer.MemoryCopy(sPtr, dPtr, count, count);
```
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Memory/Buffer2DExtensions.cs]: `Unsafe` stuff is [worse](https://github.com/SixLabors/ImageSharp/blob/026df1e4c88dec0512f8f0670fa2dee5ae59cd9e/tests/ImageSharp.Benchmarks/General/CopyBuffers.cs#L158) for large buffers, specially on classic CLR. (Alignment doesn't matter here.)

**Pull request merged**